Перевод статьи Focus management and inert с CSS tricks (#140)
web-standards-ru/web-standards.ru: src/articles/focus-and-inert/index.md

---
title: 'Управление фокусом и атрибут inert'
date: 2020-10-19
author: eric-bailey
source:
    title: 'Focus management and inert'
    url: 'https://css-tricks.com/focus-management-and-inert/'
translators:
    - mikhail-danyushin
editors:
    - vadim-makeev
layout: article.njk
tags:
    - article
    - html
    - a11y
preview: ''
featured: true
---

Множество вспомогательных технологий используют навигацию с клавиатуры для помощи в восприятии и взаимодействии с контентом. Один из способов подобной навигации — клавиша Tab. Должно быть, вы знакомы с ним, если используете Tab для быстрого перемещения между полями формы без необходимости тянуться к мышке или трекпаду.

Tab будет перемещаться по интерактивным элементам в том порядке, в котором они отображаются в DOM. Вот почему так важно, чтобы порядок исходного кода соответствовал визуальной иерархии вашего дизайна.

Список [интерактивных элементов](https://www.w3.org/TR/html52/dom.html%23interactive-content), по которым можно пройтись клавишей Tab:

> - [Ссылки](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/a), с заполненным атрибутом `href`

Reviewer: ```suggestion
- [Ссылки](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/a) с заполненным атрибутом `href`,
```
Запятая перед предлогом "с" тут не требуется, а вот добавить запятую после первого пункта списка было бы неплохо.

- [`<button>`](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/button),

> - [`<input>`](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/input) и [`<textarea>`](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/textarea), с сопутствующим им [`<label>`](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/label),

Reviewer: ```suggestion
- [`<input>`](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/input) и [`<textarea>`](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/textarea) с сопутствующим им [`<label>`](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/label),
```
Тут тоже не нужна запятая перед "с".

- [`<select>`](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/select),
- [`<details>`](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/details),

> - [`<audio>`](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/audio) и [`<video>`](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/video)при наличии контролов,

Reviewer: ```suggestion
- [`<audio>`](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/audio) и [`<video>`](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/video) при наличии контролов,
```
Пропущен пробел после ссылки.

    </video>
    <figcaption>
        Фокус следует по домашней странице. Начиная с логотипа, затем к товарам, услугам, вакансиям, блогу, контактам и останавливается на кнопке «Learn more».
    </figcaption>
</figure>

## Управление фокусом

Управлять фокусом — значит определять, какие элементы могут его получать, а какие нет.
Это один из самых сложных аспектов при разработке веб-интерфейсов, но он важен для доступности сайтов и приложений.

### Полезные практики управления фокусом

**В 99% случаев вам стоит оставить порядок фокуса в покое.** Не устану это повторять.

Состояние фокуса будет работать без дополнительных усилий, если вы используете `<button>` для кнопок, `<a>` для ссылок, `<input>` для полей форм и т. д.

В редких случаях необходимо прописать фокус элементам, не предназначенным для этого. Вот некоторые рекомендации относительно того, как реализовать это доступно и интуитивно понятно для пользователя.

#### ✅ Следует: узнать побольше про атрибут `tabindex`

`tabindex` делает элементы фокусируемыми. В качестве значения он принимает число, в зависимости от которого меняется поведение фокуса.

#### ❌ Не следует: устанавливать `tabindex="0"` там, где это не надо

> Нет необходимости устанавливать `tabindex` интерактивным элементам, которые могут получать фокус с клавиатуры (например, `<button>`). Кроме того, вам не нужно прописывать `tabindex` неинтерактивным элементам для возможности их прочтения вспомогательными устройствами (на самом деле, отсутствие `role` и доступного имени является ошибкой WCAG). Это может повлечь [нежелательный пользовательский опыт](https://adrianroselli.com/2019/02/uncanny-a11y.html#Tabindex) использования вспомогательных устройств — у которых попросту есть свои собственные способы чтения контента.

Reviewer: ```suggestion
Нет необходимости устанавливать `tabindex` для интерактивных элементов, которые могут получать фокус с клавиатуры (например, `<button>`). Кроме того, вам не нужно прописывать `tabindex` неинтерактивным элементам, чтобы их могли прочесть вспомогательные устройства (на самом деле, отсутствие `role` и доступного имени является ошибкой с точки зрения WCAG). Это может привести к [нежелательному пользовательскому опыту](https://adrianroselli.com/2019/02/uncanny-a11y.html#Tabindex) у тех, кто использует вспомогательные устройства. У таких пользователей уже есть другие, ожидаемые ими способы чтения контента.
```

> Нет необходимости устанавливать `tabindex` интерактивным элементам, которые могут получать фокус с клавиатуры (такие, как, например, `<button>`). Кроме того, вам не нужно прописывать `tabindex` неинтерактивным элементам в целях возможности их прочтения вспомогательными устройствами (на самом деле, отсутствие `role` и доступного имени является ошибкой WCAG). Это может повлечь [нежелательный пользовательский опыт](https://adrianroselli.com/2019/02/uncanny-a11y.html#Tabindex) использования вспомогательных устройств — у которых попросту есть свои собственные способы чтения контента.

Reviewer 2: ```suggestion
Нет необходимости устанавливать `tabindex` интерактивным элементам, которые могут получать фокус с клавиатуры (например, `<button>`). Кроме того, вам не нужно прописывать `tabindex` неинтерактивным элементам для возможности их прочтения вспомогательными устройствами (на самом деле, отсутствие `role` и доступного имени является ошибкой WCAG). Это может повлечь [нежелательный пользовательский опыт](https://adrianroselli.com/2019/02/uncanny-a11y.html#Tabindex) использования вспомогательных устройств — у которых попросту есть свои собственные способы чтения контента.
```

> Нет необходимости устанавливать `tabindex` для интерактивных элементов, которые могут получать фокус с клавиатуры (например, `<button>`). Кроме того, вам не нужно прописывать `tabindex` неинтерактивным элементам, чтобы их могли прочесть вспомогательные устройства (на самом деле, отсутствие `role` и доступного имени является ошибкой с точки зрения WCAG). На самом деле, это даже [усложняет навигацию](https://adrianroselli.com/2019/02/uncanny-a11y.html#Tabindex) для тех, кто использует вспомогательные устройства. У таких пользователей уже есть другие, ожидаемые ими способы чтения контента.

Reviewer 3: Выглядит как атрибут role, а тут речь роли элемента.
Оригинал:
```(in fact, this is a WCAG failure if no role and accessible name is present)```
```suggestion
Нет необходимости устанавливать `tabindex` для интерактивных элементов, которые могут получать фокус с клавиатуры (например, `<button>`). Кроме того, вам не нужно прописывать `tabindex` неинтерактивным элементам, чтобы их могли прочесть вспомогательные устройства (на самом деле, отсутствие роли и доступного имени является ошибкой с точки зрения WCAG). На самом деле, это даже [усложняет навигацию](https://adrianroselli.com/2019/02/uncanny-a11y.html#Tabindex) для тех, кто использует вспомогательные устройства. У таких пользователей уже есть другие, ожидаемые ими способы чтения контента.
```

#### ✅ Следует: устанавливать `tabindex="-1"` для фокуса с помощью JavaScript

> Атрибут `tabindex="-1"` используется для создания доступных интерактивных виджетов посредством JS. Прописав `tabindex="-1"` вы сделаете элемент фокусируемым для JS, а также по клику или тачу, но недоступным через клавишу Tab.

Reviewer 2: ```suggestion
Атрибут `tabindex="-1"` используется для создания доступных интерактивных виджетов посредством JS. Прописав `tabindex="-1"`, вы сделаете элемент фокусируемым для JS, а также по клику или тапу, но недоступным через клавишу Tab.
```

#### ❌ Не следует: использовать положительное значение `tabindex`

> Это антипаттерн. Установив положительное значение `tabindex` вы переопределите ожидемый порядок элементов для фокуса через Tab и запутаете пользователя. Сделать так один раз — уже плохо, несколько — полный кошмар. Серьезно, не надо так.

Reviewer 2: ```suggestion
Это антипаттерн. Установив положительное значение `tabindex`, вы переопределите ожидаемый порядок элементов для фокуса через Tab и запутаете пользователя. Сделать так один раз — уже плохо, несколько — полный кошмар. Серьезно, не надо так.
```

#### ❌ Не следует: создавать собственный порядок фокусировки

Пройтись по интерактивным элементам можно только в том порядке, в котором они следуют друг за другом. Не следует создавать множество `tabindex` со значением, которое увеличивается для каждого последующего элемента, в соответствии с вашим представлением о том, как пользователь должен читать ваш сайт. Позвольте DOM сделать это за вас.

> ## Захват фокуса

Reviewer: ```suggestion
## Ловушка фокуса
```
Может лучше "ловушка фокуса"?

> Иногда бывает необходимость запретить состояние фокуса. Хороший пример запрета — это [захват фокуса](https://hiddedevries.nl/en/blog/2017-01-29-using-javascript-to-trap-focus-in-an-element), временно ограничивающий его события на родительском элементе.

Reviewer: ```suggestion
Иногда есть необходимость запретить состояние фокуса. Хороший пример запрета — это [ловушка фокуса](https://hiddedevries.nl/en/blog/2017-01-29-using-javascript-to-trap-focus-in-an-element), которая временно ограничивает событие фокуса у родительского элемента и у его дочерних элементов.
```
Обычно я встречала перевод "ловушка фокуса" у этого термина. Мне он кажется более устоявшимся, чем "захват фокуса . Но могу ошибаться.

> Захват фокуса не стоит путать с ограничением [навигации с клавиатуры](https://www.w3.org/TR/UNDERSTANDING-WCAG20/keyboard-operation-trapping.html). Ограничение навигации с клавиатуры является причиной невозможности закрыть виджет или перейти к другому компоненту из-за неправильно прописанной логики.

Reviewer 3: ```suggestion
Ловушку фокуса не стоит путать с [ловушкой клавиатуры)](https://www.w3.org/TR/UNDERSTANDING-WCAG20/keyboard-operation-trapping.html). Ловушки клавиатуры — это ситуации, когда невозможно закрыть виджет или перейти к другому компоненту из-за вложенного цикла плохо прописанной логики.
```

- Для консистентности заменяем "захват" фокуса на "ловушку".
- Нет явного противопоставления терминов, я бы использовал схожие названия. И, вроде бы, в русском нет, устоявшегося термина "ограничение навигации с клавиатуры".

> Примером ситуации, когда необхдимо захватить фокус, может быть появление модального окна.

Reviewer 3: ```suggestion
Практический пример того, как вы могли бы использовать ловушку фокуса — это модальное окно:
```
Проще и ближе к переводу.

<figure>
    <video src="video/trapped-focus.mp4"
        controls loop muted playsinline>
    </video>
    <figcaption>
        Фокус проходит по странице и открывает модальное окно, чтобы продемонстрировать отмену фокуса. Далее фокус двигается в рамках контента модального окна, на кнопку «Play», кнопку «Cancel», кнопку «Purchase» и кнопку закрытия (всё это время фокус на странице заблокирован). После закрытия модального окна он возвращается к исходному положению на странице до его открытия.
    </figcaption>
</figure>

### Почему это важно?

Удержание фокуса в пределах модального окна помогает понять, что является его контентом, а что нет. Это аналогично тому, как зрячий человек может видеть, как окно всплывает поверх контента сайта или приложения. Это важная информация, если:

- У вас [плохое зрение или слепота](https://adrianroselli.com/2017/02/not-all-screen-reader-users-are-blind.html) и вы полагаетесь на скринридер, чтобы узнать об изменениях после взаимодействия со страницей;

> - У вас плохое зрение и дисплей с увеличением, при котором фокусировка за пределами модального окна может сбить с толку и дезориентировать;

Reviewer 3: "дисплей с увеличением" вызовет вопросы. Но не знаю, что это и как нормально перевести. В оригинале "magnified display".
Может речь про программную экранную лупу?

> - У вас плохое зрение и дисплей с увеличением, при котором фокусировка за пределами модального окна может сбить с толку и дезориентировать;

Reviewer 4: ```suggestion
- У вас плохое зрение и интерфейс увеличен с помощью экранной лупы, при которой фокусировка за пределами модального окна может сбить с толку и дезориентировать;
```

> - Вы перемещаетесь исключительно с помощью клавиатуры и в случае закрытия модального окна можете потеряться среди страницы в попытках к нему вернуться.

Reviewer 2: ```suggestion
- Вы перемещаетесь исключительно с помощью клавиатуры и в случае закрытия модального окна можете потеряться внутри страницы в попытках к нему вернуться.
```

### Как это сделать?

Надежно управлять фокусом — дело сложное. Нужно прибегнуть к JavaScript, чтобы:

1. Определить родительский блок для всех фокусируемых элементов на странице;

> 2. Определить границы захваченного контента (например, модального окна), включая первый и последний фокусируемый элемент;

Reviewer: ```suggestion
2. Определить границы содержимого ловушки фокуса (например, модального окна), включая первый и последний фокусируемый элемент;
```

> 3. Убрать как интерактивность, так и видимость всего, что может иметь фокус и находится вне рамок захваченного контента;

Reviewer: ```suggestion
3. Убрать как интерактивность, так и видимость всего, что может иметь фокус и находится вне рамок содержимого ловушки фокуса;
```

> 4. Переместить фокус в захваченный контент;

Reviewer: ```suggestion
4. Переместить фокус на содержимое ловушки фокуса;
```

> 5. Слушать события, сигнализирующие об уходе с выделенной области (сохрание, отмена, нажатие Esc и т. д.)

Reviewer 2: ```suggestion
5. Слушать события, сигнализирующие об уходе с выделенной области (сохранение, отмена, нажатие Esc и т. д.)
```

> 6. Уйти с захваченной области

Reviewer 2: Где-то в списке точки стоят, где-то нет. Давайте все списки к одному стилю приведём и везде проставим точки.

> 6. Уйти с захваченной области;

Reviewer 3: Не забыть изменить "захват" на "ловушку".

> 6. Уйти с захваченной области;

Reviewer: ```suggestion
6. Выйти из содержимого ловушки фокуса, когда сработает нужное событие;
```

> 5. Слушать события, сигнализирующие об уходе с выделенной области (сохранение, отмена, нажатие Esc и так далее);

Reviewer 5: ```suggestion
5. Обрабатывать события, сигнализирующие об уходе с выделенной области (сохранение, отмена, нажатие Esc и так далее);
```

7. Вернуть раннее отменённую интерактивность;

> 8. Переместить фокус обратно на интерактивный элемент, вызвавший блокировку содержимого.

Reviewer: ```suggestion
8. Переместить фокус обратно на интерактивный элемент, который вызвал ловушку фокуса.
```

### Зачем нам это?

> Не стану врать: все эти действия отнимают много времени. Но всё же, управление фокусом и удобный порядок фокусировки являются частью [WCAG (руководства по доступности в вебе)](https://www.w3.org/TR/UNDERSTANDING-WCAG20/navigation-mechanisms-focus-order.html). Тема достаточна важна, чтобы считать её частью [международного правового стандарта об удобстве пользования](https://www.w3.org/WAI/policies/).

Reviewer: ```suggestion
Не стану врать: все эти действия отнимают много времени. Но всё же, управление фокусом и удобный порядок фокусировки являются частью [WCAG (руководства по доступности в вебе)](https://www.w3.org/TR/UNDERSTANDING-WCAG20/navigation-mechanisms-focus-order.html). Тема достаточна важна, чтобы считать её частью [международного правового стандарта о юзабилити](https://www.w3.org/WAI/policies/).
```
Кажется, что в данном случае уместнее не переводить и использовать "юзабилити", потому что юзабилити — более широкое понятие, чем просто удобство использования.

> Не стану врать: все эти действия отнимают много времени. Но всё же, управление фокусом и удобный порядок фокусировки являются частью [WCAG (руководства по доступности в вебе)](https://www.w3.org/TR/UNDERSTANDING-WCAG20/navigation-mechanisms-focus-order.html). Тема достаточна важна, чтобы считать её частью [международного правового стандарта о юзабилити](https://www.w3.org/WAI/policies/).

Reviewer 3: Есть распространённый перевод названия гайда, я бы использовал его.
```suggestion
Не стану врать: все эти действия отнимают много времени. Но всё же, управление фокусом и удобный порядок фокусировки являются частью [WCAG (руководства по обеспечению доступности веб-контента)](https://www.w3.org/TR/UNDERSTANDING-WCAG20/navigation-mechanisms-focus-order.html). Тема достаточна важна, чтобы считать её частью [международного правового стандарта о юзабилити](https://www.w3.org/WAI/policies/).
```

### Видимость

Существует небольшой трюк, с помощью которого можно легко ограничить видимость и интерактивность элемента.

> [Скринридеры обладают режимом взаимодействия](https://tink.uk/understanding-screen-reader-interaction-modes/), который позволяет им проходить по странице виртуальным курсором. А ещё виртуальный курсор позволяет пользователю скринридера обнаруживать неинтерактивные части страницы (заголовки, списки и т. д.). В отличие от использования Tab, виртуальный курсор доступен только пользователям скринридера.

Reviewer 2: Вижу здесь Ё. Давайте тоже на странице её использование сделаем однообразным: либо используем, либо нет, либо там, где без точечек над Е слово не понятно без контекста.

Reviewer 4: Политика с «Ё» на сайте простая: буква используется всегда.

> [Скринридеры обладают режимом взаимодействия](https://tink.uk/understanding-screen-reader-interaction-modes/), который позволяет им проходить по странице виртуальным курсором. А ещё виртуальный курсор позволяет пользователю скринридера обнаруживать неинтерактивные части страницы (заголовки, списки и т. д.). В отличие от использования Tab, виртуальный курсор доступен только пользователям скринридера.

Reviewer: ```suggestion
[У скринридеров есть режим взаимодействия](https://tink.uk/understanding-screen-reader-interaction-modes/), который позволяет им проходить по странице или просматривать её с помощью виртуального курсора. А ещё виртуальный курсор позволяет пользователю скринридера обнаруживать неинтерактивные части страницы (заголовки, списки и т.  д.). В отличие от использования Tab, виртуальный курсор доступен только пользователям скринридера.
```

> При управлении фокусом может потребоваться ограничить возможность обнаружения содержимого виртуальным курсором. В нашем примере с модальным окном это значит предотвратить пользователя от случайного выхода за рамки окна при чтении контента.

Reviewer 2: ```suggestion
При управлении фокусом может потребоваться ограничить возможность обнаружения содержимого виртуальным курсором. В нашем примере с модальным окном это значит предотвратить случайный выход пользователя за рамки окна при чтении контента.
```

Видимость может быть заблокирована с помощью правильного применения атрибута [`aria-hidden="true"`](https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/ARIA_Techniques/Using_the_aria-hidden_attribute). А вот с интерактивностью есть один тонкий нюанс.

## Знакомство с `inert`

> [Атрибут `inert`](https://whatpr.org/html/4288/interaction.html#the-inert-attribute) — [глобальный атрибут в HTML](https://developer.mozilla.org/en-US/docs/Web/HTML/Global_attributes), позволяющий легко убирать и восстанавливать видимость интерактивных элементов, а так же их взможность получать состояние фокуса. Вот пример:

Reviewer 2: ```suggestion
[Атрибут `inert`](https://whatpr.org/html/4288/interaction.html#the-inert-attribute) — [глобальный атрибут в HTML](https://developer.mozilla.org/en-US/docs/Web/HTML/Global_attributes), позволяющий легко убирать и восстанавливать видимость интерактивных элементов, а также их возможность получать состояние фокуса. Вот пример:
```

```html
<body>
    <div
        aria-labelledby="modal-title"
        class="c-modal"
        id="modal"
        role="dialog"
        tabindex="-1">
        <div role="document">
            <h2 id="modal-title">Save changes?</h2>

>             <p>The changes you have made will be lost if you do not save them.<p>

Reviewer 2: Может, и тут переведём текст интерфейса?

Reviewer 4: К сожалению, он не переведён на видео, так что я бы оставил

            <button type="button">Save</button>
            <button type="button">Discard</button>
        </div>
    </div>
    <main inert>
        <!-- ... -->
    </main>
</body>
```

> Я намеренно избегаю использования `<dialog>`для модального окна из-за [многочисленных проблем с поддержкой](https://www.scottohara.me/blog/2019/03/05/open-dialog.html).

Reviewer: ```suggestion
Я намеренно избегаю использования `<dialog>` для модального окна из-за [многочисленных проблем с поддержкой](https://www.scottohara.me/blog/2019/03/05/open-dialog.html).
```

Атрибут `inert` задан элементу `<main>`, следующему сразу после модального окна сохранения изменений. Теперь всё содержимое `<main>` не может попасть в фокус и по нему нельзя кликнуть.

> Фокус ограничен рамками модального окна. Когда мы закрываем его, мы убираем `inert` из `<main>`. Этот способ ограничения фокуса проще, чем другие методы.

Reviewer: ```suggestion
Фокус ограничен рамками модального окна. Когда мы закрываем его, то убираем `inert` из `<main>`. Этот способ управления ловушкой фокуса проще, чем другие методы.
```

> **Помните:** Событие закрытия модального окна может быть вызванно не только нажатием соотетствующих кнопок, но и нажатием на Esc. А еще некоторые модальные окна могут быть закрыты по клику на оверлей.

Reviewer 2: ```suggestion
**Помните:** событие закрытия модального окна может быть вызвано не только нажатием соответствующих кнопок, но и нажатием Esc. А еще некоторые модальные окна могут быть закрыты по клику на оверлей.
```

> **Помните:** событие закрытия модального окна может быть вызвано не только нажатием соответствующих кнопок, но и нажатием Esc. А ещё некоторые модальные окна могут быть закрыты по клику на оверлей.

Reviewer: ```suggestion
**Помните:** событие закрытия модального окна может быть вызвано не только нажатием на кнопки внутри нашего модального окна из примера, но и при нажатии на Esc. А ещё некоторые модальные окна могут быть закрыты по клику на оверлей.
```

### Поддержка `inert`

> Все последние версии [Edge, Chrome, and Opera поддерживают `inert`](https://caniuse.com/#feat=mdn-api_htmlelement_inert) если [включены экспериментальные функции веб-платформ](https://css-irl.info/how-to-enable-experimental-web-platform-features/). Firefox будет поддерживать [в скором времени](https://bugzilla.mozilla.org/show_bug.cgi?id=1655722)! Единственное исключение — Safari. Как мобильный, так и десктопный.

Reviewer 2: ```suggestion
Все последние версии [Edge, Chrome и Opera поддерживают `inert`](https://caniuse.com/#feat=mdn-api_htmlelement_inert), если [включены экспериментальные функции веб-платформ](https://css-irl.info/how-to-enable-experimental-web-platform-features/). Firefox будет поддерживать [в скором времени](https://bugzilla.mozilla.org/show_bug.cgi?id=1655722)! Единственное исключение — Safari. Как мобильный, так и десктопный.
```

> Все последние версии [Edge, Chrome и Opera поддерживают `inert`](https://caniuse.com/#feat=mdn-api_htmlelement_inert), если [включены экспериментальные функции веб-платформ](https://css-irl.info/how-to-enable-experimental-web-platform-features/). Firefox будет поддерживать [в скором времени](https://bugzilla.mozilla.org/show_bug.cgi?id=1655722)! Единственное исключение — Safari. Как мобильный, так и десктопный.

Reviewer: ```suggestion
Все последние версии [Edge, Chrome и Opera поддерживают `inert`](https://caniuse.com/#feat=mdn-api_htmlelement_inert), если [включены экспериментальные функции веб-платформ](https://css-irl.info/how-to-enable-experimental-web-platform-features/). Также [скоро появится](https://bugzilla.mozilla.org/show_bug.cgi?id=1655722) его поддержка в Firefox! Единственное исключение — Safari, как мобильный, так и десктопный.
```

> Мне бы очень хотелось, чтобы Apple реализовала встроенную поддержку `inert`. Хотя и [существует полифил](https://github.com/WICG/inert), у него [серьёзные проблемы с поддержкой скринридерами](https://www.scottohara.me/blog/2019/03/05/open-dialog.html#does-the-polyfill-help). Нехорошо!

Reviewer: ```suggestion
Мне бы очень хотелось, чтобы Apple реализовала встроенную поддержку `inert`. Хоть [полифил и существует](https://github.com/WICG/inert), но у него [серьёзные проблемы с поддержкой скринридерами](https://www.scottohara.me/blog/2019/03/05/open-dialog.html#does-the-polyfill-help). Нехорошо!
```

> Вдобавок, хочется обратить внимание на [запись в ридми полифила](https://github.com/WICG/inert#performance-and-gotchas):

Reviewer 2: ```suggestion
Вдобавок хочется обратить внимание на [пометку в README полифила](https://github.com/WICG/inert#performance-and-gotchas):
```

> > Полифил обойдется вам дорого с точки зрения производительности в отличие от нативного `inert`, т. к. он требует изрядное количество перемещений по DOM.

Reviewer 2: ```suggestion
> Полифил обойдется вам дорого с точки зрения производительности в отличие от нативного `inert`, т. к. он требует изрядного количества перемещений по DOM.
```

> > Полифил обойдётся вам дорого с точки зрения производительности в отличие от нативного `inert`, т. к. он требует изрядного количества перемещений по DOM.

Reviewer: ```suggestion
> Полифил обойдётся вам дорого с точки зрения производительности, в отличие от нативного `inert`, т. к. он требует изрядного количества перемещений по DOM.
```

> Перемещение по DOM подразумевает что JavaSсript в полифиле требует высокой вычислительной мощности и, следовательно, в конечном итоге замедлит пользователя.

Reviewer 2: ```suggestion
Перемещение по DOM подразумевает, что JavaSсript в полифиле требует высокой вычислительной мощности и, следовательно, в конечном итоге замедлит пользователя.
```

> Перемещение по DOM подразумевает, что JavaSсript в полифиле требует высокой вычислительной мощности и, следовательно, в конечном итоге замедлит пользователя.

Reviewer 5: Русское "с"

```suggestion
Перемещение по DOM подразумевает, что JavaScript в полифиле требует высокой вычислительной мощности и, следовательно, в конечном итоге замедлит пользователя.
```

> Перемещение по DOM подразумевает, что JavaScript в полифиле требует высокой вычислительной мощности и, следовательно, в конечном итоге замедлит пользователя.

Reviewer 3: Чуть ближе к сути, сам пользователь не будет замедлен.
```suggestion
Перемещение по DOM подразумевает, что JavaScript в полифиле требует высокой вычислительной мощности и, следовательно, в конечном итоге замедлит использование.
```

> Для устройств с низким энергопотреблением таких как [бюджетные Android смартфоны и устаревшие ноутбуки](https://www.pewresearch.org/fact-tank/2019/05/07/digital-divide-persists-even-as-lower-income-americans-make-gains-in-tech-adoption/ft\_17-03-21\_low-incometech_smartphone/), а также тех, что выполняют сложные задачи (например, запуск нескольких приложений сразу), это может привести к зависанию или сбоям. Нативная браузерная поддержка делает такие процессы менее затратными в этом плане, так как у браузера есть доступ ко всем частям DOM, в отличии от JS.

Reviewer 2: ```suggestion
Для устройств с низким энергопотреблением, таких как [бюджетные Android смартфоны и устаревшие ноутбуки](https://www.pewresearch.org/fact-tank/2019/05/07/digital-divide-persists-even-as-lower-income-americans-make-gains-in-tech-adoption/ft\_17-03-21\_low-incometech_smartphone/), а также тех, что выполняют сложные задачи (например, запуск нескольких приложений сразу), это может привести к зависанию или сбоям. Нативная браузерная поддержка делает такие процессы менее затратными в этом плане, так как у браузера есть доступ ко всем частям DOM, в отличие от JS.
```

> Для устройств с низким энергопотреблением, таких как [бюджетные Android смартфоны и устаревшие ноутбуки](https://www.pewresearch.org/fact-tank/2019/05/07/digital-divide-persists-even-as-lower-income-americans-make-gains-in-tech-adoption/ft\_17-03-21\_low-incometech_smartphone/), а также тех, что выполняют сложные задачи (например, запуск нескольких приложений сразу), это может привести к зависанию или сбоям. Нативная браузерная поддержка делает такие процессы менее затратными в этом плане, так как у браузера есть доступ ко всем частям DOM, в отличие от JS.

Reviewer: ```suggestion
Для устройств с низким энергопотреблением, таких как [бюджетные Android-смартфоны и устаревшие ноутбуки](https://www.pewresearch.org/fact-tank/2019/05/07/digital-divide-persists-even-as-lower-income-americans-make-gains-in-tech-adoption/ft\_17-03-21\_low-incometech_smartphone/), а также тех, что выполняют сложные задачи (например, запуск нескольких приложений сразу), это может привести к зависанию или сбоям. Нативная браузерная поддержка делает такие процессы менее затратными в этом плане, так как у браузера есть доступ ко всем частям DOM, в отличие от JS.
```

#### Safari

> Лично я разочарован тем, что Apple не поддерживает `inert`. Хотя я понимаю, что добавление новых функций в браузер — невероятно сложная и трудная работа, `inert` кажется функцией, которую Apple могла бы поддерживать гораздо раньше.

Reviewer: ```suggestion
Лично я разочарован тем, что Apple не поддерживает `inert`. Хотя понимаю, что добавление новых функций в браузер — невероятно сложная и трудная работа. `inert` кажется функцией, которую Apple стоило бы начать поддерживать гораздо раньше.
```

> macOS и iOS всегда имели хорошую поддержку доступности, а поддержка вспомогательных технологий — важная часть [их маркетинговой кампании](https://www.apple.com/accessibility/). Поддержка `inert` представляется как естественное продолжение миссии Apple, ввиду того, что сама эта функция смогла бы облегчить разработку доступных веб-интерфейсов в разы.

Reviewer: ```suggestion
macOS и iOS всегда имели хорошую поддержку доступности, а поддержка вспомогательных технологий — важная часть [их маркетинговой кампании](https://www.apple.com/accessibility/). Поддержка `inert` представляется как естественное продолжение миссии Apple ввиду того, что сама эта функция смогла бы облегчить разработку доступных веб-интерфейсов в разы.
```

> К сожалению, Apple держит в тайне, когда воявится поддержка этого атрибута. Поэтому поддержка `inert` — все еще открытый вопрос.

Reviewer 2: ```suggestion
К сожалению, Apple держит в тайне, когда появится поддержка этого атрибута. Поэтому поддержка `inert` — все еще открытый вопрос.
```

> К сожалению, Apple держит в тайне, когда появится поддержка этого атрибута. Поэтому поддержка `inert` — все еще открытый вопрос.

Reviewer: ```suggestion
К сожалению, Apple держит в тайне, когда появится поддержка этого атрибута. Поэтому поддержка `inert` — всё ещё открытый вопрос.
```

#### Igalia

[Igalia](https://www.igalia.com/) — компания, работающая над функциями браузеров. Сейчас они проводят эксперимент, в котором каждый может проголосовать за те возможности браузеров, которые ему хотелось бы видеть. Конечно, моя статья совсем не про это, но вы можете [узнать больше в Smashing Magazine](https://www.smashingmagazine.com/2020/07/crowdfunding-web-platform-features-open-prioritization/).

> Одно из предложений Igalia — это [добавление поддержки `inert` в WebKit](https://www.igalia.com/open-prioritization/#inertwebkit). Если вы искали возможность принять участие в улучшении доступности веба, но не знали с чего начать, я могу вам [ее предоставить](https://opencollective.com/html-inert-in-webkit-safari). 5 $, 10 $, 25 $ — совсем необязательно жертвовать большие суммы, даже маленький вклад ценен.

Reviewer 2: ```suggestion
Одно из предложений Igalia — это [добавление поддержки `inert` в WebKit](https://www.igalia.com/open-prioritization/#inertwebkit). Если вы искали возможность принять участие в улучшении доступности веба, но не знали с чего начать, я могу вам [ее предоставить](https://opencollective.com/html-inert-in-webkit-safari). $5, $10, $25 — совсем необязательно жертвовать большие суммы, даже маленький вклад ценен.
```

Reviewer 4: $5, $10, $25 — эта запись принята в англоязычной типографике. Располагать символ валюты в зависимости от валюты — это плохой подход. Ты же не будешь писать ₽100?

Reviewer 2: Я тут полностью согласен с Бирманом: https://ilyabirman.ru/meanwhile/all/dollar-sign/
Мы у себя в одном крупном проекте используем именно международные обозначения валют. Рубли по международным правилам пишутся после количества.

Reviewer 4: И всё же, у нас принято так: 100 €, 100 $, 100 ₽

[Пожертвовать](https://opencollective.com/html-inert-in-webkit-safari)

## Итог

> Управление фокусом требует навыка и осторожности. Но это того стоит. Атрибут `inert` облегчит это.

Reviewer: ```suggestion
Управление фокусом требует некоторых навыков и осторожности, но это того стоит. Атрибут `inert` значительно облегчит эту задачу.
```

> Такие техники, как `inert` являются примером одной из самых сильных сторон веба: [способность кодифицировать неожиданное поведение](https://www.w3.org/TR/html-design-principles/#pave-the-cowpaths) в нечто простое и эффективное.

Reviewer 2: ```suggestion
Такие техники, как `inert`, являются примером одной из самых сильных сторон веба: [способности кодифицировать неожиданное поведение](https://www.w3.org/TR/html-design-principles/#pave-the-cowpaths) в нечто простое и эффективное.
```

### Что почитать

> - [Controlling focus with tabindex](https://youtu.be/Pe0Ce1WtnUM) (A11ycasts, эпизод 4)
> - [Using the tabindex attribute](https://developer.paciellogroup.com/blog/2014/08/using-the-tabindex-attribute/) (The Paciello Group)
> - [Using JavaScript to trap focus in an element](https://hiddedevries.nl/en/blog/2017-01-29-using-javascript-to-trap-focus-in-an-element) (Хидде де Врис)

Reviewer: ```suggestion
- [Controlling focus with tabindex](https://youtu.be/Pe0Ce1WtnUM) (A11ycasts, эпизод 4).
- [Using the tabindex attribute](https://developer.paciellogroup.com/blog/2014/08/using-the-tabindex-attribute/) (The Paciello Group).
- [Using JavaScript to trap focus in an element](https://hiddedevries.nl/en/blog/2017-01-29-using-javascript-to-trap-focus-in-an-element) (Хидде де Врис).
```

---
title: 'Управление фокусом и атрибут inert'
date: 2020-10-19
author: eric-bailey
source:
    title: 'Focus management and inert'
    url: 'https://css-tricks.com/focus-management-and-inert/'
translators:
    - mikhail-danyushin
editors:
    - vadim-makeev
layout: article.njk
tags:
    - article
    - html
    - a11y

> preview: ''

Reviewer 6: Превью бы написать.

Reviewer 4: Это последний шаг, надо сначала зафиналить

Author: А кто обычно пишет превью? Я бы мог попробовать

Reviewer 4: Ревью, это что-то вроде тизера к статье. Чаще собирается из первого или несколькольких первых абзацев. Его имеет смысл писать перед публикацией, когда эти абзацы уже закончены.

- [Ссылки](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/a), с заполненным атрибутом `href`
- [`<button>`](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/button),
- [`<input>`](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/input) и `[<textarea>](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/textarea)`, с сопутствующим им [label](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/label),
- [`<select>`](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/select),
- [`<details>`](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/details),
- [`<audio>`](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/audio) и [`<video>`](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/video) при наличии контролов,
- [`<object>`](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/object), в зависимости от того, как он используется,
- любой элемент с `overflow: scroll` в Firefox,
- любой элемент с атрибутом [`contenteditable`](https://developer.mozilla.org/en-US/docs/Web/HTML/Global_attributes/contenteditable), и

> - любой элемент с установленным `[tabindex](https://developer.mozilla.org/en-US/docs/Web/HTML/Global_attributes/tabindex)` (о нём чуть позже).

Reviewer 6: @pepelsbey почему-то выше в бектики обернул только теги, а тут всю ссылку целиком?)

Reviewer 4: Не доделал :)
